connmgr/connmanager_test.go

```diff
@@ -0,0 +1,373 @@
+		MaxOutbound:   1,
+		Dial:          mockDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+		OnDisconnection: func(c *connmgr.ConnReq) {
+			disconnected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	go cmgr.Connect(cr)
+	cmgr.Start()
+	gotConnReq := <-connected
+	wantID := cr.ID()
+	gotID := gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState := cr.State()
+	wantState := connmgr.ConnEstablished
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+
+	cmgr.Disconnect(cr.ID())
+	gotConnReq = <-disconnected
+	wantID = cr.ID()
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnDisconnected
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+
+	gotConnReq = <-connected
+	wantID = cr.ID()
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnEstablished
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+
+	cmgr.Remove(cr.ID())
+	gotConnReq = <-disconnected
+	wantID = cr.ID()
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnDisconnected
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+	cmgr.Stop()
+}
+
+// TestMaxRetryDuration tests the maximum retry duration.
+//
+// We have a timed dialer which initially returns err but after RetryDuration
+// hits maxRetryDuration returns a mock conn.
+func TestMaxRetryDuration(t *testing.T) {
+	networkUp := make(chan struct{})
+	time.AfterFunc(5*time.Millisecond, func() {
+		close(networkUp)
+	})
+	timedDialer := func(network, address string) (net.Conn, error) {
+		select {
+		case <-networkUp:
+			return mockDialer(network, address)
+		default:
+			return nil, errors.New("network down")
+		}
+	}
+
+	connected := make(chan *connmgr.ConnReq)
+	cmgr, err := connmgr.New(&connmgr.Config{
+		RetryDuration: time.Millisecond,
+		MaxOutbound:   1,
+		Dial:          timedDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	go cmgr.Connect(cr)
+	cmgr.Start()
+	// retry in 1ms
+	// retry in 2ms - max retry duration reached
+	// retry in 2ms - timedDialer returns mockDial
+	<-connected
```

Comment:
What if it doesn't work as expected?  The text will never fail here.  Need a select statement with a timeout longer than `timedDailer` with a `t.Fatal` branch.